tao_compiler/mlir/disc/transforms/disc_dot_merge.cc

```diff
@@ -0,0 +1,557 @@
+      }
+      const auto& dimension_numbers = dotShape.dimension_numbers;
+      const auto& lhs_batch_dims = dimension_numbers.getLhsBatchingDimensions();
+      const auto& rhs_batch_dims = dimension_numbers.getRhsBatchingDimensions();
+      const auto& lhs_contracting_dims =
+          dimension_numbers.getLhsContractingDimensions();
+      const auto& rhs_contracting_dims =
+          dimension_numbers.getRhsContractingDimensions();
+      hash = llvm::hash_combine(hash,
+                                llvm::hash_combine_range(lhs_batch_dims.begin(),
+                                                         lhs_batch_dims.end()));
+      hash = llvm::hash_combine(hash,
+                                llvm::hash_combine_range(rhs_batch_dims.begin(),
+                                                         rhs_batch_dims.end()));
+      hash = llvm::hash_combine(
+          hash, llvm::hash_combine_range(lhs_contracting_dims.begin(),
+                                         lhs_contracting_dims.end()));
+      hash = llvm::hash_combine(
+          hash, llvm::hash_combine_range(rhs_contracting_dims.begin(),
+                                         rhs_contracting_dims.end()));
+      hash = llvm::hash_combine(hash, mlir::hash_value(dotShape.element_type));
+      return hash;
+    }
+  };
+
+  struct BatchCluster {
+    BatchCluster(Operation* op, int op_id) : leader_op_id(op_id) {
+      ops.push_back(op);
+    }
+
+    // Merges `other` into this cluster, and clears `other`.
+    void merge(BatchCluster& other) {
+      ops.insert(ops.end(), other.ops.begin(), other.ops.end());
+      other.ops.clear();
+    }
+
+    // ID of the representative node of this cluster.
+    int leader_op_id;
+
+    // Dot ops to be batched.
+    SmallVector<Operation*> ops;
+  };
+
+  bool buildShapedDotMap();
+  bool batchingDots();
+  Value expandDim0(OpBuilder& builder, Location& loc, Value value);
+  bool applyBatching(BatchCluster& cluster);
+
+ private:
+  FuncOp func_;
+  // A map to help to cluster dot that have same shape and dim-numbers together.
+  std::unordered_map<DotShape, SmallVector<mhlo::DotGeneralOp>, DotShapeHash>
+      equal_shape_dot_map_;
+};
+
+void DotBatchingConverter::run() {
+  if (!buildShapedDotMap()) {
+    return;
+  }
+  batchingDots();
+}
+
+bool DotBatchingConverter::buildShapedDotMap() {
+  ShapeAnalysis analysis(func_);
+  if (failed(analysis.run())) {
+    LLVM_DEBUG(llvm::dbgs() << "ShapeAnalysis failes for dot merge.\n");
+    return false;
+  }
+  func_.walk([&](mhlo::DotGeneralOp op) {
+    DotShape dot_shape;
+    Value lhs = op.lhs();
+    Value rhs = op.rhs();
+    // Initialize `dimension_numbers`.
+    dot_shape.dimension_numbers = op.dot_dimension_numbers();
+    auto lhs_batch_dims =
+        dot_shape.dimension_numbers.getLhsBatchingDimensions();
+    auto rhs_batch_dims =
+        dot_shape.dimension_numbers.getRhsBatchingDimensions();
+    assert(lhs_batch_dims.size() == rhs_batch_dims.size());
```

Comment:
do we need to check the actual value of lhs_batch_dims and rhs_batch_dims are same.

Reply:
We plan to add the same-operand-dot-merging process just before the batching process.

Reply:
I mean we need to check the contents of lhs/rhs batch_dims are same, right?

Reply:
This should be guaranteed in previous passes, if the contents of lhs/rhs batch_dims do not equal, the dot general is illegal.